eth/tracers/taiko_api.go

```diff
@@ -0,0 +1,434 @@
+package tracers
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/internal/ethapi"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethereum/go-ethereum/trie"
+)
+
+type TaikoBackend interface {
+	BlockChain() *core.BlockChain
+}
+
+// provingPreflightResult is the result of a proving preflight request.
+type provingPreflightResult struct {
+	Block             *types.Block                   `json:"block"`
+	InitAccountProofs []*ethapi.AccountResult        `json:"initAccountProofs"`
+	Contracts         map[common.Hash]*hexutil.Bytes `json:"contracts"`
+	AncestorHashes    map[uint64]common.Hash         `json:"ancestorHashes"`
+	Error             string                         `json:"error,omitempty"`
+}
+
+// provingPreflightTask represents a single block preflight task.
+type provingPreflightTask struct {
+	statedb   *state.StateDB          // Intermediate state prepped for preflighting
+	parent    *types.Block            // Parent block of the block to preflight
+	block     *types.Block            // Block to preflight the transactions from
+	release   StateReleaseFunc        // The function to release the held resource for this task
+	preflight *provingPreflightResult // Preflight results produced by the task
+}
+
+// ProvingPreflights traces the blockchain and gets preflights from the start block to the end block
+// and returns a subscription to the results. This function is intended to be
+// used with long-running operations, as tracing a chain can be time-consuming.
+//
+// Parameters:
+// - ctx: The context for the operation.
+// - start: The starting block number for the trace.
+// - end: The ending block number for the trace.
+// - config: The configuration for the trace.
+//
+// Returns:
+// - *rpc.Subscription: A subscription to the trace results.
+// - error: An error if the operation fails, or if the end block is not after the start block.
+//
+// Note:
+//   - The function requires a notifier to be present in the context, otherwise it
+//     returns an error indicating that notifications are unsupported.
+func (api *API) ProvingPreflights(ctx context.Context, start, end rpc.BlockNumber, config *TraceConfig) (*rpc.Subscription, error) {
+	if start == 0 {
+		return nil, fmt.Errorf("start block must be greater than 0")
+	}
+	// Adjust the start block to the parent one
+	start -= 1
```

Comment:
`-1` means the range start from parent number?

Reply:
Yes, [start, end]